lf-queue/lfq.c

```diff
@@ -0,0 +1,214 @@
+#include <assert.h>
+#include <errno.h>
+#include <stdbool.h>
+
+#include "atomics.h"
+#include "lfq.h"
+
+#define MAX_FREE 150
+
+static bool in_hp(struct lfq_ctx *ctx, struct lfq_node *node)
+{
+    for (int i = 0; i < ctx->MAX_HP_SIZE; i++) {
+        if (atomic_load(&ctx->HP[i]) == node)
+            return true;
+    }
+    return false;
+}
+
+/* add to tail of the free list */
+static void insert_pool(struct lfq_ctx *ctx, struct lfq_node *node)
+{
+    atomic_store(&node->free_next, NULL);
+    struct lfq_node *old_tail = XCHG(&ctx->fpt, node); /* seq_cst */
+    atomic_store(&old_tail->free_next, node);
+}
+static void free_pool(struct lfq_ctx *ctx, bool freeall)
+{
+    bool old = 0;
+    if (!CAS(&ctx->is_freeing, &old, 1))
+        return;
+
+    for (int i = 0; i < MAX_FREE || freeall; i++) {
+        struct lfq_node *p = ctx->fph;
+        if ((!atomic_load(&p->can_free)) || (!atomic_load(&p->free_next)) ||
+            in_hp(ctx, (struct lfq_node *) p))
+            goto final;
```

Comment:
Instead of using `goto`, can you simply use `break` statement?

Reply:
Done!